File: sdks/javascript/src/resilience/tracing.ts
/**
 * OpenTelemetry Tracing Integration for Resilience Patterns.
 * Provides tracing spans for all resilience patterns.
 * @module aether/resilience/tracing
 */

import { TracingConfig, TracingContext } from './types';

// Check if OpenTelemetry is available
let TRACING_AVAILABLE = false;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let apiModule: any = null;

try {
  // Dynamic import check - will be undefined if not installed
  // @ts-ignore - optional dependency
  apiModule = require('@opentelemetry/api');
  TRACING_AVAILABLE = apiModule !== null;
} catch {
  // OpenTelemetry not installed
}

/**
 * Check if OpenTelemetry tracing is available.
 *
 * @returns `true` if the `@opentelemetry/api` package is installed and loaded.
 */
export function isTracingAvailable(): boolean {
  return TRACING_AVAILABLE;
}

/**
 * Get a tracer instance for resilience patterns.
 *
 * @param serviceName - The service name to report in traces
 *                      (default: `'aether-resilience'`).
 * @returns A tracer instance, or `null` if tracing is not available.
 */
export function getTracer(serviceName = 'aether-resilience'): unknown {
  if (!TRACING_AVAILABLE || !apiModule) {
    return null;
  }

  try {
    const trace = apiModule.trace;
    return trace.getTracer(serviceName);
  } catch {
    return null;
  }
}

/**
 * Minimal span interface for tracing operations.
 *
 * Provides attribute-setting, event-adding, and status-setting capabilities
 * compatible with OpenTelemetry spans.
 */
interface Span {
  /** Set a key-value attribute on the span. */
  setAttribute(key: string, value: unknown): void;
  /** Record an event with optional attributes. */
  addEvent(name: string, attributes?: Record<string, unknown>): void;
  /** Set the span status. */
  setStatus(status: { code: number; message?: string }): void;
  /** End the span. */
  end(): void;
}

/**
 * No-op span implementation used when tracing is disabled.
 *
 * All methods are safe no-ops that return `this` for chaining.
 * @internal
 */
class NoOpSpan implements Span {
  setAttribute(_key: string, _value: unknown): this {
    return this;
  }
  addEvent(_name: string, _attributes?: Record<string, unknown>): this {
    return this;
  }
  setStatus(_status: { code: number; message?: string }): this {
    return this;
  }
  end(): void {}
}

const noOpSpan = new NoOpSpan();

/**
 * Execute a function within a tracing span.
 *
 * If OpenTelemetry is available, creates a real span; otherwise uses a
 * no-op span. The span status is set to OK on success or ERROR on failure.
 *
 * @typeParam T - The return type of the function.
 * @param spanName - Name for the tracing span.
 * @param fn       - Function receiving the span and returning a promise.
 * @returns The function result.
 *
 * @example
 * ```typescript
 * const result = await withTracing('my_operation', async (span) => {
 *   span.setAttribute('my.key', 'value');
 *   return doWork();
 * });
 * ```
 */
export async function withTracing<T>(
  spanName: string,
  fn: (span: Span) => Promise<T>
): Promise<T> {
  if (!TRACING_AVAILABLE || !apiModule) {
    return fn(noOpSpan);
  }

  const tracer = getTracer();
  if (!tracer) {
    return fn(noOpSpan);
  }

  try {
    const trace = apiModule.trace;
    const context = apiModule.context;

    // Start span
    const span = (tracer as any).startActiveSpan(
      spanName,
      (span: Span) => {
        // Execute function with span
        return fn(span).then(
          (result) => {
            span.setStatus({ code: 0 }); // OK
            span.end();
            return result;
          },
          (error) => {
            span.setStatus({ code: 2, message: error?.message }); // ERROR
            span.end();
            throw error;
          }
        );
      }
    );

    return span;
  } catch {
    // If tracing fails, just run the function
    return fn(noOpSpan);
  }
}

// ============================================
// Traced Wrappers
// ============================================

/**
 * Wrap a circuit breaker operation with tracing.
 *
 * @typeParam T - The return type.
 * @param name  - The circuit breaker name.
 * @param state - The current circuit state.
 * @param fn    - The async function to trace.
 * @returns The function result.
 */
export function tracedCircuitBreaker<T>(
  name: string,
  state: string,
  fn: () => Promise<T>
): Promise<T> {
  return withTracing(`circuit_breaker.${name}.execute`, async (span) => {
    span.setAttribute('circuit_breaker.name', name);
    span.setAttribute('circuit_breaker.state', state);

    try {
      const result = await fn();
      span.setAttribute('circuit_breaker.result', 'success');
      return result;
    } catch (error) {
      span.setAttribute('circuit_breaker.result', 'error');
      throw error;
    }
  });
}

/**
 * Wrap a retry operation with tracing.
 *
 * @typeParam T - The return type.
 * @param name       - The retry policy name.
 * @param attempt    - The current attempt number (1-based).
 * @param maxAttempts - The maximum number of attempts.
 * @param fn         - The async function to trace.
 * @returns The function result.
 */
export function tracedRetry<T>(
  name: string,
  attempt: number,
  maxAttempts: number,
  fn: () => Promise<T>
): Promise<T> {
  return withTracing(`retry.${name}.attempt`, async (span) => {
    span.setAttribute('retry.name', name);
    span.setAttribute('retry.attempt', attempt);
    span.setAttribute('retry.max_attempts', maxAttempts);

    try {
      const result = await fn();
      span.setAttribute('retry.result', 'success');
      return result;
    } catch (error) {
      span.setAttribute('retry.result', 'error');
      span.addEvent('retry.failed', {
        'error.message': error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  });
}

/**
 * Wrap a rate limiter operation with tracing.
 *
 * @typeParam T - The return type.
 * @param name    - The rate limiter name.
 * @param allowed - Whether the request was allowed.
 * @param fn      - The async function to trace.
 * @returns The function result.
 */
export function tracedRateLimiter<T>(
  name: string,
  allowed: boolean,
  fn: () => Promise<T>
): Promise<T> {
  return withTracing(`rate_limiter.${name}.execute`, async (span) => {
    span.setAttribute('rate_limiter.name', name);
    span.setAttribute('rate_limiter.allowed', allowed);

    try {
      const result = await fn();
      span.setAttribute('rate_limiter.result', 'success');
      return result;
    } catch (error) {
      span.setAttribute('rate_limiter.result', 'error');
      throw error;
    }
  });
}

/**
 * Wrap a bulkhead operation with tracing.
 *
 * @typeParam T - The return type.
 * @param name           - The bulkhead name.
 * @param active         - Number of currently active calls.
 * @param maxConcurrent  - Maximum allowed concurrent calls.
 * @param fn             - The async function to trace.
 * @returns The function result.
 */
export function tracedBulkhead<T>(
  name: string,
  active: number,
  maxConcurrent: number,
  fn: () => Promise<T>
): Promise<T> {
  return withTracing(`bulkhead.${name}.execute`, async (span) => {
    span.setAttribute('bulkhead.name', name);
    span.setAttribute('bulkhead.active', active);
    span.setAttribute('bulkhead.max_concurrent', maxConcurrent);

    try {
      const result = await fn();
      span.setAttribute('bulkhead.result', 'success');
      return result;
    } catch (error) {
      span.setAttribute('bulkhead.result', 'error');
      throw error;
    }
  });
}

/**
 * Wrap a health check operation with tracing.
 *
 * @typeParam T - The return type.
 * @param name      - The health checker name.
 * @param checkName - The individual check name.
 * @param fn        - The async function to trace.
 * @returns The function result.
 */
export function tracedHealthCheck<T>(
  name: string,
  checkName: string,
  fn: () => Promise<T>
): Promise<T> {
  return withTracing(`health_check.${name}.${checkName}`, async (span) => {
    span.setAttribute('health_check.name', name);
    span.setAttribute('health_check.check', checkName);

    try {
      const result = await fn();
      span.setAttribute('health_check.result', 'success');
      return result;
    } catch (error) {
      span.setAttribute('health_check.result', 'error');
      throw error;
    }
  });
}

// ============================================
// Instrumentation Helper
// ============================================

/**
 * Resilience instrumentation configuration and span factory.
 *
 * Provides convenience methods for creating typed spans for each
 * resilience pattern without requiring direct OpenTelemetry usage.
 *
 * @example
 * ```typescript
 * const instrumentation = new ResilienceInstrumentation({
 *   serviceName: 'my-service',
 *   sampleRate: 0.5,
 * });
 *
 * const { span, end } = instrumentation.circuitBreaker('api', 'closed');
 * // ... do work ...
 * end(); // or end(error)
 * ```
 */
export class ResilienceInstrumentation {
  private tracer: unknown;
  private config: TracingConfig;

  /**
   * Create a new ResilienceInstrumentation.
   *
   * @param config - Partial tracing configuration.
   */
  constructor(config: Partial<TracingConfig> = {}) {
    this.config = {
      enabled: config.enabled ?? true,
      serviceName: config.serviceName ?? 'aether-resilience',
      sampleRate: config.sampleRate ?? 1.0,
    };

    if (this.config.enabled) {
      this.tracer = getTracer(this.config.serviceName);
    }
  }

  /**
   * Create a circuit breaker span and end callback.
   *
   * @param name  - The circuit breaker name.
   * @param state - The current circuit state.
   * @returns An object with the span and an `end` function.
   */
  circuitBreaker(name: string, state: string): {
    span: Span;
    end: (error?: Error) => void;
  } {
    const span = noOpSpan;
    return {
      span,
      end: (error?: Error) => {
        if (error) {
          span.setAttribute('error', error.message);
        }
        span.end();
      },
    };
  }

  /**
   * Create a retry span and end callback.
   *
   * @param name        - The retry policy name.
   * @param attempt     - The current attempt number.
   * @param maxAttempts - The maximum number of attempts.
   * @returns An object with the span and an `end` function.
   */
  retry(name: string, attempt: number, maxAttempts: number): {
    span: Span;
    end: (error?: Error) => void;
  } {
    const span = noOpSpan;
    return {
      span,
      end: (error?: Error) => {
        if (error) {
          span.setAttribute('error', error.message);
        }
        span.end();
      },
    };
  }

  /**
   * Create a rate limiter span and end callback.
   *
   * @param name    - The rate limiter name.
   * @param allowed - Whether the request was allowed.
   * @returns An object with the span and an `end` function.
   */
  rateLimiter(name: string, allowed: boolean): {
    span: Span;
    end: (error?: Error) => void;
  } {
    const span = noOpSpan;
    return {
      span,
      end: (error?: Error) => {
        if (error) {
          span.setAttribute('error', error.message);
        }
        span.end();
      },
    };
  }

  /**
   * Create a bulkhead span and end callback.
   *
   * @param name          - The bulkhead name.
   * @param active        - Number of currently active calls.
   * @param maxConcurrent - Maximum allowed concurrent calls.
   * @returns An object with the span and an `end` function.
   */
  bulkhead(name: string, active: number, maxConcurrent: number): {
    span: Span;
    end: (error?: Error) => void;
  } {
    const span = noOpSpan;
    return {
      span,
      end: (error?: Error) => {
        if (error) {
          span.setAttribute('error', error.message);
        }
        span.end();
      },
    };
  }
}

// ============================================
// Exports
// ============================================

export {
  TRACING_AVAILABLE,
};

export type { Span };

File: sdks/python/aether_sdk/resilience/tracing.py
"""
OpenTelemetry Tracing Integration for Resilience Patterns

Provides tracing spans for all resilience patterns to integrate with OpenTelemetry.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional, Dict, Any, Callable
from contextlib import contextmanager
import functools
import time

# Check if OpenTelemetry is available
TRACING_AVAILABLE = False
_tracer_module = None

try:
    from opentelemetry import trace as _tracer_module
    from opentelemetry.trace import SpanKind, Status, StatusCode
    TRACING_AVAILABLE = True
except ImportError:
    pass


class TracingContext:
    """Context manager for tracing spans."""
    
    def __init__(
        self,
        tracer: Optional[Any],
        span_name: str,
        attributes: Optional[Dict[str, Any]] = None,
    ):
        self.tracer = tracer
        self.span_name = span_name
        self.attributes = attributes or {}
        self.span = None
        self.start_time: Optional[float] = None
    
    def __enter__(self):
        if not TRACING_AVAILABLE or not self.tracer:
            return self
        
        # Start span
        self.span = self.tracer.start_span(
            self.span_name,
            kind=SpanKind.INTERNAL if TRACING_AVAILABLE else None,
            attributes=self.attributes,
        )
        self.start_time = time.time()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self.span:
            return
        
        if TRACING_AVAILABLE:
            if exc_type:
                self.span.set_status(Status(StatusCode.ERROR))
                self.span.set_attribute("error.type", exc_type.__name__)
                self.span.set_attribute("error.message", str(exc_val))
            else:
                self.span.set_status(Status(StatusCode.OK))
            
            if self.start_time:
                self.span.set_attribute(
                    "duration_ms",
                    int((time.time() - self.start_time) * 1000)
                )
            self.span.end()
    
    def set_attribute(self, key: str, value: Any) -> None:
        if self.span:
            self.span.set_attribute(key, value)
    
    def add_event(self, name: str, attributes: Optional[Dict[str, Any]] = None) -> None:
        if self.span:
            self.span.add_event(name, attributes=attributes or {})


def get_tracer(service_name: str = "aether-resilience") -> Optional[Any]:
    """Get or create a tracer for resilience patterns."""
    if not TRACING_AVAILABLE or not _tracer_module:
        return None
    
    try:
        tracer_provider = _tracer_module.get_tracer_provider()
        return tracer_provider.get_tracer(service_name)
    except Exception:
        return None


# ============================================
# Traced Decorators
# ============================================

def traced_circuit_breaker(name: str = "default"):
    """Decorator to add tracing to circuit breaker operations."""
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(self, *args, **kwargs):
            if not TRACING_AVAILABLE:
                return await func(self, *args, **kwargs)
            
            tracer = get_tracer()
            if not tracer:
                return await func(self, *args, **kwargs)
            
            state_val = self.state.value if hasattr(self, 'state') else "unknown"
            with TracingContext(
                tracer,
                f"circuit_breaker.{name}.{func.__name__}",
                {
                    "circuit_breaker.name": name,
                    "circuit_breaker.state": state_val,
                },
            ) as ctx:
                try:
                    result = await func(self, *args, **kwargs)
                    ctx.set_attribute("circuit_breaker.result", "success")
                    return result
                except Exception as e:
                    ctx.set_attribute("circuit_breaker.result", "rejected")
                    raise
        
        return wrapper
    return decorator


def traced_retry(name: str = "default"):
    """Decorator to add tracing to retry operations."""
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(self, *args, **kwargs):
            if not TRACING_AVAILABLE:
                return await func(self, *args, **kwargs)
            
            tracer = get_tracer()
            if not tracer:
                return await func(self, *args, **kwargs)
            
            max_attempts = self._config.max_attempts if hasattr(self, '_config') else 0
            with TracingContext(
                tracer,
                f"retry.{name}.{func.__name__}",
                {
                    "retry.name": name,
                    "retry.max_attempts": max_attempts,
                },
            ) as ctx:
                try:
                    result = await func(self, *args, **kwargs)
                    ctx.set_attribute("retry.result", "success")
                    return result
                except Exception as e:
                    ctx.set_attribute("retry.result", "exhausted")
                    raise
        
        return wrapper
    return decorator


def traced_rate_limiter(name: str = "default"):
    """Decorator to add tracing to rate limiter operations."""
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(self, *args, **kwargs):
            if not TRACING_AVAILABLE:
                return await func(self, *args, **kwargs)
            
            tracer = get_tracer()
            if not tracer:
                return await func(self, *args, **kwargs)
            
            with TracingContext(
                tracer,
                f"rate_limiter.{name}.{func.__name__}",
                {
                    "rate_limiter.name": name,
                },
            ) as ctx:
                try:
                    result = await func(self, *args, **kwargs)
                    if hasattr(result, 'allowed'):
                        ctx.set_attribute("rate_limiter.allowed", result.allowed)
                    return result
                except Exception as e:
                    ctx.set_attribute("rate_limiter.allowed", False)
                    raise
        
        return wrapper
    return decorator


def traced_bulkhead(name: str = "default"):
    """Decorator to add tracing to bulkhead operations."""
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(self, *args, **kwargs):
            if not TRACING_AVAILABLE:
                return await func(self, *args, **kwargs)
            
            tracer = get_tracer()
            if not tracer:
                return await func(self, *args, **kwargs)
            
            with TracingContext(
                tracer,
                f"bulkhead.{name}.{func.__name__}",
                {
                    "bulkhead.name": name,
                },
            ) as ctx:
                try:
                    result = await func(self, *args, **kwargs)
                    ctx.set_attribute("bulkhead.result", "success")
                    return result
                except Exception as e:
                    ctx.set_attribute("bulkhead.result", "rejected")
                    raise
        
        return wrapper
    return decorator


# ============================================
# Helper Functions
# ============================================

def create_resilience_span(
    operation: str,
    pattern_type: str,
    pattern_name: str,
    attributes: Optional[Dict[str, Any]] = None,
) -> Optional[TracingContext]:
    """Create a tracing span for resilience operations."""
    if not TRACING_AVAILABLE:
        return None
    
    tracer = get_tracer()
    if not tracer:
        return None
    
    return TracingContext(
        tracer,
        f"{pattern_type}.{pattern_name}.{operation}",
        attributes or {},
    )


def record_resilience_event(
    pattern_type: str,
    event_name: str,
    attributes: Optional[Dict[str, Any]] = None,
) -> None:
    """Record a resilience event in the current span."""
    if not TRACING_AVAILABLE or not _tracer_module:
        return
    
    try:
        span = _tracer_module.get_current_span()
        if span:
            span.add_event(
                f"{pattern_type}.{event_name}",
                attributes=attributes or {},
            )
    except Exception:
        pass


def set_resilience_attribute(key: str, value: Any) -> None:
    """Set an attribute on the current span."""
    if not TRACING_AVAILABLE or not _tracer_module:
        return
    
    try:
        span = _tracer_module.get_current_span()
        if span:
            span.set_attribute(key, value)
    except Exception:
        pass


# ============================================
# Instrumentation Class
# ============================================

class ResilienceInstrumentation:
    """Provides instrumentation utilities for resilience patterns."""
    
    def __init__(self, service_name: str = "aether-resilience"):
        self.tracer = get_tracer(service_name)
    
    @contextmanager
    def trace_circuit_breaker(
        self,
        name: str,
        state: str,
        operation: str,
    ):
        """Create a circuit breaker trace context."""
        with TracingContext(
            self.tracer,
            f"circuit_breaker.{name}.{operation}",
            {
                "circuit_breaker.name": name,
                "circuit_breaker.state": state,
            },
        ) as ctx:
            yield ctx
    
    @contextmanager
    def trace_retry(
        self,
        name: str,
        attempt: int,
        max_attempts: int,
        operation: str,
    ):
        """Create a retry trace context."""
        with TracingContext(
            self.tracer,
            f"retry.{name}.{operation}",
            {
                "retry.name": name,
                "retry.attempt": attempt,
                "retry.max_attempts": max_attempts,
            },
        ) as ctx:
            yield ctx
    
    @contextmanager
    def trace_rate_limiter(
        self,
        name: str,
        operation: str,
        requests_per_second: int = 0,
    ):
        """Create a rate limiter trace context."""
        with TracingContext(
            self.tracer,
            f"rate_limiter.{name}.{operation}",
            {
                "rate_limiter.name": name,
                "rate_limiter.requests_per_second": requests_per_second,
            },
        ) as ctx:
            yield ctx
    
    @contextmanager
    def trace_bulkhead(
        self,
        name: str,
        operation: str,
        active: int = 0,
        max_concurrent: int = 0,
    ):
        """Create a bulkhead trace context."""
        with TracingContext(
            self.tracer,
            f"bulkhead.{name}.{operation}",
            {
                "bulkhead.name": name,
                "bulkhead.active": active,
                "bulkhead.max_concurrent": max_concurrent,
            },
        ) as ctx:
            yield ctx
    
    @contextmanager
    def trace_health_check(
        self,
        name: str,
        check_name: str,
    ):
        """Create a health check trace context."""
        with TracingContext(
            self.tracer,
            f"health_check.{name}.{check_name}",
            {
                "health_check.name": name,
                "health_check.check": check_name,
            },
        ) as ctx:
            yield ctx


# ============================================
# Exports
# ============================================

__all__ = [
    'TRACING_AVAILABLE',
    'TracingContext',
    'get_tracer',
    'traced_circuit_breaker',
    'traced_retry',
    'traced_rate_limiter',
    'traced_bulkhead',
    'create_resilience_span',
    'record_resilience_event',
    'set_resilience_attribute',
    'ResilienceInstrumentation',
]
